network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use channel::libra_channel;
+use futures::{sink::SinkExt, StreamExt};
+use libra_config::config::RoleType;
+use libra_logger::prelude::*;
+use libra_metrics::{register_histogram, DurationHistogram};
+use libra_network_address::NetworkAddress;
+use libra_types::on_chain_config::{OnChainConfigPayload, ValidatorSet};
+use network::{
+    common::NetworkPublicKeys,
+    connectivity_manager::{ConnectivityRequest, DiscoverySource},
+};
+use once_cell::sync::Lazy;
+use std::{convert::TryFrom, time::Instant};
+
+/// Histogram of idle time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_IDLE_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_idle_duration_s",
+            "Histogram of idle time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Histogram of busy time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_busy_duration_s",
+            "Histogram of busy time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+    // Collect the set of address updates.
+    let updates: Vec<ConnectivityRequest> = validator_keys
+        .into_iter()
+        .map(|validator| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                validator.account_address().clone(),
+                vec![
+                    NetworkAddress::try_from(&validator.config().validator_network_address)
+                        .expect("WTF"),
```

Comment:
I'm torn.  On the one hand your instinct is much safer in production.  On the other hand, this is a value which is coming out of the config and should have been properly structured when it went in;  so if this try_from fails there are much bigger problems afoot....

Reply:
If it is coming from the blockchain, then I would actually expect it to be fully vetted and parseable *before* the blockchain accepts the change including the new address.  The fact that it is on the blockchain is actually the only reason I would consier trusting that the data is parsable.

(Admittedly, such expectations are dangerous and double checking is always at least as safe as relying on a previously performed check)

Reply:
the `validator_network_address` on chain is just an opaque byte array (`vector<u8>` in move). there are several reasons for not parsing them further on chain:

1. we want to be able to upgrade network transport and handshake protocols w/o needing to do an upgrade of the consensus-critical on-chain validator set. instead, validators can just advertise new and old protocols simultaneously.

2. clients running an older binary might not understand some of the advertised addresses and will need to ignore them, which implies that we have to parse the addresses anyway. this is also a natural occurrence during an upgrade rollout.

3. move lang is not really mature enough yet to implement the parsing code. for example, i don't think it supports enums, among other things.

4. additional maintenance burden of support move parsing logic and ensuring it matches the rust logic.

so this code should probably just ignore network addresses it can't understand. we should probably log those that fail, however.